fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexMaintainer.java

```diff
@@ -311,10 +318,51 @@ void deleteDocument(Tuple groupingKey, Integer partitionId, Tuple primaryKey) th
 
     @Override
     public CompletableFuture<Void> mergeIndex() {
-        return partitioner.rebalancePartitions()
+        return rebalancePartitions()
                 .thenCompose(ignored -> directoryManager.mergeIndex(partitioner, indexAnalyzerSelector.provideIndexAnalyzer("")));
     }
 
+    @Nonnull
+    @VisibleForTesting
+    public LucenePartitioner getPartitioner() {
+        return partitioner;
+    }
+
+    @SuppressWarnings("PMD.CloseResource") // the runner is closed in a whenComplete on the future returned
+    public CompletableFuture<Void> rebalancePartitions() {
+        if (!partitioner.isPartitioningEnabled()) {
+            return CompletableFuture.completedFuture(null);
+        }
+        final FDBRecordStore.Builder storeBuilder = state.store.asBuilder();
+        FDBDatabaseRunner runner = state.context.newRunner();
+        final Index index = state.index;
+        return rebalancePartitions(runner, storeBuilder, index)
+                .whenComplete((result, error) -> {
+                    runner.close();
+                });
+    }
+
+    private static CompletableFuture<Void> rebalancePartitions(final FDBDatabaseRunner runner, final FDBRecordStore.Builder storeBuilder, final Index index) {
+        AtomicReference<RecordCursorContinuation> continuation = new AtomicReference<>(RecordCursorStartContinuation.START);
+        return AsyncUtil.whileTrue(() -> {
+            return runner.runAsync(context -> {
+                return storeBuilder.setContext(context).openAsync().thenCompose(store -> {
+                    store.getIndexDeferredMaintenanceControl().setAutoMergeDuringCommit(false);
+                    return ((LuceneIndexMaintainer)store.getIndexMaintainer(index))
```

Comment:
If the store changes between transactions, and the index no longer is a `LuceneIndexMantainer` this will fail with a `ClassCastException`